Admin: Adds setting to disable creating initial admin user

shavonn opened the pull request:
Closes #19038

Commit:
added disable_admin_user to disable creation of default user

marefr commented on pkg/services/sqlstore/sqlstore.go:
```diff
@@ -294,7 +294,7 @@ type ITestDB interface {
 	t.Helper()
 	sqlstore := &SqlStore{}
-	sqlstore.skipEnsureAdmin = true
+	sqlstore.skipEnsureAdmin = setting.DisableAdminUser
```
This is only used for test and `skipEnsureAdmin` should always be false (as before). You want to read the `setting.DisableAdminUser` in the `Init()` method since that method is called when Grafana starts up.

shavonn replied:
I just figured that out! LOL Thanks!

marefr reviewed (commented):
Added a comment

Commit:
check setting in init

marefr commented on pkg/services/sqlstore/sqlstore.go:
```diff
@@ -294,7 +296,7 @@ type ITestDB interface {
 	t.Helper()
 	sqlstore := &SqlStore{}
-	sqlstore.skipEnsureAdmin = true
+	sqlstore.skipEnsureAdmin = false
```
Think the test fails because you'll need to set `setting.DisableAdminUser` to true for the `sqlstore.Cfg` a couple of lines below.

shavonn replied:
It was supposed to be true. I meant to change it back and put the wrong val.

Commit:
sqlstore.skipEnsureAdmin back to true

marefr commented on pkg/services/sqlstore/sqlstore.go:
```diff
@@ -99,6 +99,8 @@ func (ss *SqlStore) Init() error {
 	// Register handlers
 	ss.addUserQueryAndCommandHandlers()
 
+	ss.skipEnsureAdmin = setting.DisableAdminUser
```
You shouldn't use the global setting here. Use `ss.skipEnsureAdmin = ss.Cfg.DisableAdminUser`.

marefr replied:
You probably need to change code in setting package and move `DisableAdminUser` to "cfg" struct.

marefr replied:
Think you can remove `ss.skipEnsureAdmin` alltogether and check only the `DisableAdminUser` setting.

marefr commented on pkg/setting/setting.go:
```diff
@@ -755,6 +756,7 @@ func (cfg *Cfg) Load(args *CommandLineArgs) error {
 	}
 
 	// admin
+	DisableAdminUser = security.Key("disable_admin_user").MustBool(false)
```
Here, instead of using public variable `DisableAdminUser`, assign it to `cfg.DisableAdminUser` and move the variable `DisableAdminUser` into `Cfg` struct.

Commit:
move option to config

Commit:
remove skipEnsureAdmin

olivierlemasle commented:
I think there is another issue.

Currently, the default organization "Main Org." is created along with this admin user. In a Grafana installation with LDAP mapping configured (use case of #19038) and `disable_admin_user = true`, the default organization is not created during Grafana startup.

I just tried to test this pull request with LDAP, and when the LDAP user tries to log in, an error occurs because the organization does not exist:

```
DBUG[09-30|17:15:37] LDAP users found                         logger=ldap users="([]*models.ExternalUserInfo) (len=1 cap=1) {\n (*models.ExternalUserInfo)(0xc0001ec360)({\n  OAuthToken: (*oauth2.Token)(<nil>),\n  AuthModule: (string) (len=4) \"ldap\",\n  AuthId: (string) (len=40) \"cn=ldap-admin,ou=users,dc=grafana,dc=org\",\n  UserId: (int64) 0,\n  Email: (string) \"\",\n  Login: (string) (len=10) \"ldap-admin\",\n  Name: (string) (len=10) \"ldap-admin\",\n  Groups: ([]string) (len=1 cap=1) {\n   (string) (len=37) \"cn=admins,ou=groups,dc=grafana,dc=org\"\n  },\n  OrgRoles: (map[int64]models.RoleType) (len=1) {\n   (int64) 1: (models.RoleType) (len=5) \"Admin\"\n  },\n  IsGrafanaAdmin: (*bool)(<nil>),\n  IsDisabled: (bool) false\n })\n}\n"
EROR[09-30|17:15:37] Error while trying to authenticate user  logger=context userId=0 orgId=0 uname= error="user does not belong to org"
EROR[09-30|17:15:37] Request Completed                        logger=context userId=0 orgId=0 uname= method=POST path=/login status=500 remote_addr=127.0.0.1 time_ms=29 size=91 referer=http://localhost:3000/login

```

marefr commented:
@olivierlemasle thanks. We'll lock into this. Probably need to create the main org separately.

olivierlemasle commented:
Yes, I guess a function `ensureMainOrg()` would be useful, just before

```go
if ss.Cfg.DisableAdminUser {
    return nil
}
```

Commit:
move creation of main org

torkelo commented on conf/defaults.ini:
```diff
@@ -155,6 +155,9 @@ google_tag_manager_id =
 #################################### Security ############################
 [security]
+# disable creation of admin user on first start of grafana
+disable_admin_user = false
```
I would prefer a name that is a bit more clear, disable_initial_admin_creation , or something that make it clear that this not disabling any user but the creation of the first default admin user

marefr replied:
Good point

Commit:
variable name changes

torkelo commented:
can you rebase with master or investigate the CI test errors

marefr commented:
Sorry that reviewing this has stalled. At first thought this was going to be quite straight forward, but it wasn't due to that creating the (first) main org and admin user is baked into `CreateUserCommand` handler.

My suggestion is to not use the `CreateUserCommand` for this specific use case and implement all needed sql in `EnsureAdminUser` for creating either only the main org or either both main org and admin user. Doing this we can wrap everything in a transaction.

Questions: 
- Currently if there exist an admin user the user/org is not created. What happens when we disable default admin account and there are no other admin account (until ldap sync) and Grafana starts up?
  - Probably need to check if main org exists separately from if admin user exists

marefr committed:
Merge remote-tracking branch 'origin/master' into 19038-disable-admin-user

marefr committed:
fix after merge

marefr committed:
Merge remote-tracking branch 'origin/master' into 19038-disable-admin-user

marefr committed:
various fixes

Fix so that default main org and admin user are not created in all tests.
Try handle when disable creating initial admin user to create main org.

marefr committed:
remove unused code

marefr committed:
only create admin user/main if there are no users in the system

Fixes so that the condition for creating default admin user/main org
is same as before this feature.

torkelo commented on pkg/services/sqlstore/org.go:
```diff
@@ -214,3 +217,60 @@ func DeleteOrg(cmd *m.DeleteOrgCommand) error {
+			return org.Id, nil
+		}
+		if setting.AutoAssignOrgId == 1 {
+			org.Name = "Main Org."
```
should this be orgName?

marefr replied:
In this PR I would say no, it currently works this way in master. Need to look into this later.

torkelo reviewed (approved):
Looks good, just one minor thing

marefr committed:
const for main org

marefr committed:
Merge remote-tracking branch 'origin/master' into 19038-disable-admin-user

marefr committed:
docs: disable_initial_admin_creation

torkelo reviewed (approved):
Looks good

marefr reviewed (approved):
LGTM
